components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java

```diff
@@ -194,6 +197,103 @@ public boolean isAttachedLabel(String user, String labelId) throws APIManagement
         return false;
     }
 
+    @Override
+    public List<Environment> getAllEnvironments(String tenantDomain) throws APIManagementException {
+        List<Environment> dynamicEnvs = apiMgtDAO.getAllEnvironments(tenantDomain);
+        // gateway environment name should be unique, ignore environments defined in api-manager.xml with the same name
+        // if a dynamic (saved in database) environment exists.
+        List<String> dynamicEnvNames = dynamicEnvs.stream().map(Environment::getName).collect(Collectors.toList());
+        List<Environment> allEnvs = new ArrayList<>(dynamicEnvs.size() + APIUtil.getReadOnlyEnvironments().size());
+        // add read only environments first and dynamic environments later
+        APIUtil.getReadOnlyEnvironments().values().stream().filter(env -> !dynamicEnvNames.contains(env.getName())).forEach(allEnvs::add);
+        allEnvs.addAll(dynamicEnvs);
+        return allEnvs;
+    }
+
+    @Override
+    public Environment getEnvironment(String tenantDomain, String uuid) throws APIManagementException {
+        Environment env = apiMgtDAO.getEnvironment(tenantDomain, uuid);
+        if (env == null) {
+            // name is the UUID of environments configured in api-manager.xml
+            env = APIUtil.getReadOnlyEnvironments().get(uuid);
```

Comment:
better to put this logic upward.

Reply:
given priority for configured environments over dynamic environments
fixed with https://github.com/wso2/carbon-apimgt/pull/9890/commits/8b56851663554865bad62d7ac549af4a1f58f55b